File: examples/lights.rs
use processing_glfw::GlfwContext;

use bevy::math::Vec3;
use processing::prelude::*;
use processing_render::render::command::DrawCommand;

fn main() {
    match sketch() {
        Ok(_) => {
            eprintln!("Sketch completed successfully");
            exit(0).unwrap();
        }
        Err(e) => {
            eprintln!("Sketch error: {:?}", e);
            exit(1).unwrap();
        }
    };
}

fn sketch() -> error::Result<()> {
    let mut glfw_ctx = GlfwContext::new(400, 400)?;
    init(Config::default())?;

    let width = 400;
    let height = 400;
    let surface = glfw_ctx.create_surface(width, height)?;
    let graphics = graphics_create(surface, width, height, TextureFormat::Rgba16Float)?;
    let box_geo = geometry_box(100.0, 100.0, 100.0)?;
    let pbr_mat = material_create_pbr()?;
    material_set(pbr_mat, "roughness", material::MaterialValue::Float(0.0))?;

    // We will only declare lights in `setup`
    // rather than calling some sort of `light()` method inside of `draw`

    // create a directional light
    let _dir_light =
        light_create_directional(graphics, bevy::color::Color::srgb(0.35, 0.25, 0.5), 500.0)?;

    // create a point light
    let point_light_a = light_create_point(
        graphics,
        bevy::color::Color::srgb(1.0, 0.5, 0.25),
        1_000_000.0,
        200.0,
        0.5,
    )?;
    transform_set_position(point_light_a, Vec3::new(-25.0, 5.0, 51.0))?;
    transform_look_at(point_light_a, Vec3::new(0.0, 0.0, 0.0))?;

    // create another point light
    let point_light_b = light_create_point(
        graphics,
        bevy::color::Color::srgb(0.0, 0.5, 0.75),
        2_000_000.0,
        200.0,
        0.25,
    )?;
    transform_set_position(point_light_b, Vec3::new(0.0, 5.0, 50.5))?;
    transform_look_at(point_light_b, Vec3::new(0.0, 0.0, 0.0))?;

    // and a spot light, too!
    let spot_light = light_create_spot(
        graphics,
        bevy::color::Color::srgb(0.25, 0.8, 0.19),
        15.0 * 1_000_000.0,
        200.0,
        0.84,
        0.0,
        core::f32::consts::FRAC_PI_4,
    )?;
    transform_set_position(spot_light, Vec3::new(40.0, 0.0, 70.0))?;
    transform_look_at(spot_light, Vec3::new(0.0, 0.0, 0.0))?;

    graphics_mode_3d(graphics)?;
    transform_set_position(graphics, Vec3::new(100.0, 100.0, 300.0))?;
    transform_look_at(graphics, Vec3::new(0.0, 0.0, 0.0))?;

    let mut angle = 0.0;

    while glfw_ctx.poll_events() {
        graphics_begin_draw(graphics)?;

        graphics_record_command(
            graphics,
            DrawCommand::BackgroundColor(bevy::color::Color::srgb(0.18, 0.20, 0.15)),
        )?;

        graphics_record_command(graphics, DrawCommand::Fill(bevy::color::Color::WHITE))?;
        graphics_record_command(graphics, DrawCommand::Material(pbr_mat))?;

        graphics_record_command(graphics, DrawCommand::PushMatrix)?;
        graphics_record_command(graphics, DrawCommand::Rotate { angle })?;
        graphics_record_command(graphics, DrawCommand::Geometry(box_geo))?;
        graphics_record_command(graphics, DrawCommand::PopMatrix)?;

        graphics_end_draw(graphics)?;

        angle += 0.02;
    }
    Ok(())
}

File: crates/processing_pyo3/examples/lights.py
from mewnala import *

angle = 0.0

def setup():
    size(800, 600)
    mode_3d()

    # Directional Light
    dir_light = directional_light((0.5, 0.24, 1.0), 1500.0)

    # Point Lights
    point_light_a = point_light((1.0, 0.5, 0.25), 1000000.0, 200.0, 0.5)
    point_light_a.position(-25.0, 5.0, 51.0)
    point_light_a.look_at(0.0, 0.0, 0.0)

    point_light_b = point_light((0.0, 0.5, 0.75), 2000000.0, 200.0, 0.25)
    point_light_b.position(0.0, 5.0, 50.5)
    point_light_b.look_at(0.0, 0.0, 0.0)

    # Spot Light
    spot_light_a = spot_light((0.25, 0.8, 0.19), 15.0 * 1000000.0, 200.0, 0.84, 0.0, 0.7854)
    spot_light_a.position(40.0, 0.0, 70.0)
    spot_light_a.look_at(0.0, 0.0, 0.0)

def draw():
    global angle
    camera_position(100.0, 100.0, 300.0)
    camera_look_at(0.0, 0.0, 0.0)
    background(220)

    roughness(0.0)
    push_matrix()
    rotate(angle)
    box(100.0, 100.0, 100.0)
    pop_matrix()

    angle += 0.02


# TODO: this should happen implicitly on module load somehow
run()

